### DISCUSSION TITLE
Restore problem

### DISCUSSION THREAD

**Discussion by pochacho99** [teslamate-org/teslamate#2972] [Q&A] [answered]
Hello, my Docker for Windows wouldn't start, so had to re-install but lost all my containers in the process. Created new Teslamate container and tried to run the restore, but get the following output. Nothing is restored, although backup file looks to be in order. Any suggestions please?

C:\docker\teslamate>docker-compose exec -T database psql -U teslamate -d teslamate < teslamate.bck
SET
SET
SET
SET
SET
 set_config
------------

(1 row)

SET
SET
SET
SET
CREATE EXTENSION
COMMENT
CREATE EXTENSION
COMMENT
ERROR:  type "billing_type" already exists
ALTER TYPE
ERROR:  type "range" already exists
ALTER TYPE
ERROR:  type "states_status" already exists
ALTER TYPE
ERROR:  type "unit_of_length" already exists
ALTER TYPE
ERROR:  type "unit_of_temperature" already exists
ALTER TYPE
ERROR:  function "convert_celsius" already exists with same argument types
ALTER FUNCTION
ERROR:  function "convert_km" already exists with same argument types
ALTER FUNCTION
ERROR:  function "convert_m" already exists with same argument types
ALTER FUNCTION
ERROR:  function "convert_tire_pressure" already exists with same argument types
ALTER FUNCTION
SET
SET
ERROR:  relation "addresses" already exists
ALTER TABLE
ERROR:  relation "addresses_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ERROR:  relation "car_settings" already exists
ALTER TABLE
ERROR:  relation "car_settings_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ERROR:  relation "cars" already exists
ALTER TABLE
ERROR:  relation "cars_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ERROR:  relation "charges" already exists
ALTER TABLE
ERROR:  relation "charges_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ERROR:  relation "charging_processes" already exists
ALTER TABLE
ERROR:  relation "charging_processes_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ERROR:  relation "drives" already exists
ALTER TABLE
ERROR:  relation "geofences" already exists
ALTER TABLE
ERROR:  relation "geofences_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ERROR:  relation "positions" already exists
ALTER TABLE
ERROR:  relation "positions_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ERROR:  relation "schema_migrations" already exists
ALTER TABLE
ERROR:  relation "settings" already exists
ALTER TABLE
ERROR:  relation "settings_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ERROR:  relation "states" already exists
ALTER TABLE
ERROR:  relation "states_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ERROR:  relation "tokens" already exists
ALTER TABLE
ERROR:  relation "tokens_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ERROR:  relation "trips_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ERROR:  relation "updates" already exists
ALTER TABLE
ERROR:  relation "updates_id_seq" already exists
ALTER TABLE
ALTER SEQUENCE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ERROR:  duplicate key value violates unique constraint "addresses_osm_id_osm_type_index"
DETAIL:  Key (osm_id, osm_type)=(142285725, way) already exists.
CONTEXT:  COPY addresses, line 1
ERROR:  duplicate key value violates unique constraint "car_settings_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY car_settings, line 1
ERROR:  duplicate key value violates unique constraint "cars_eid_index"
DETAIL:  Key (eid)=(1493067908319761) already exists.
CONTEXT:  COPY cars, line 1
ERROR:  insert or update on table "charges" violates foreign key constraint "charges_charging_process_id_fkey"
DETAIL:  Key (charging_process_id)=(1) is not present in table "charging_processes".
ERROR:  insert or update on table "charging_processes" violates foreign key constraint "charging_processes_position_id_fkey"
DETAIL:  Key (position_id)=(843872) is not present in table "positions".
ERROR:  insert or update on table "drives" violates foreign key constraint "drives_start_position_id_fkey"
DETAIL:  Key (start_position_id)=(1225058) is not present in table "positions".
ERROR:  duplicate key value violates unique constraint "geofences_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY geofences, line 1
ERROR:  insert or update on table "positions" violates foreign key constraint "positions_drive_id_fkey"
DETAIL:  Key (drive_id)=(1) is not present in table "drives".
ERROR:  duplicate key value violates unique constraint "schema_migrations_pkey"
DETAIL:  Key (version)=(20190330150000) already exists.
CONTEXT:  COPY schema_migrations, line 1
ERROR:  duplicate key value violates unique constraint "settings_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY settings, line 1
ERROR:  duplicate key value violates unique constraint "states_pkey"
DETAIL:  Key (id)=(70) already exists.
CONTEXT:  COPY states, line 1
ERROR:  duplicate key value violates unique constraint "tokens_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY tokens, line 1
ERROR:  duplicate key value violates unique constraint "updates_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY updates, line 1
 setval
--------
    316
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
  24543
(1 row)

 setval
--------
     98
(1 row)

 setval
--------
      6
(1 row)

 setval
---------
 2620091
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
   1312
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
   1422
(1 row)

 setval
--------
     28
(1 row)

ERROR:  multiple primary keys for table "addresses" are not allowed
ERROR:  multiple primary keys for table "car_settings" are not allowed
ERROR:  multiple primary keys for table "cars" are not allowed
ERROR:  multiple primary keys for table "charges" are not allowed
ERROR:  multiple primary keys for table "charging_processes" are not allowed
ERROR:  multiple primary keys for table "geofences" are not allowed
ERROR:  multiple primary keys for table "positions" are not allowed
ERROR:  multiple primary keys for table "schema_migrations" are not allowed
ERROR:  multiple primary keys for table "settings" are not allowed
ERROR:  multiple primary keys for table "states" are not allowed
ERROR:  multiple primary keys for table "tokens" are not allowed
ERROR:  multiple primary keys for table "drives" are not allowed
ERROR:  multiple primary keys for table "updates" are not allowed
ERROR:  relation "addresses_osm_id_osm_type_index" already exists
ERROR:  relation "cars_eid_index" already exists
ERROR:  relation "cars_settings_id_index" already exists
ERROR:  relation "cars_vid_index" already exists
ERROR:  relation "cars_vin_index" already exists
ERROR:  relation "charges_charging_process_id_index" already exists
ERROR:  relation "charges_date_index" already exists
ERROR:  relation "charging_processes_address_id_index" already exists
ERROR:  relation "charging_processes_car_id_index" already exists
ERROR:  relation "charging_processes_position_id_index" already exists
ERROR:  relation "drives_end_geofence_id_index" already exists
ERROR:  relation "drives_end_position_id_index" already exists
ERROR:  relation "drives_start_geofence_id_index" already exists
ERROR:  relation "drives_start_position_id_index" already exists
ERROR:  relation "positions_car_id_index" already exists
ERROR:  relation "positions_date_index" already exists
ERROR:  relation "positions_drive_id_index" already exists
ERROR:  relation "states_car_id__end_date_IS_NULL_index" already exists
ERROR:  relation "states_car_id_index" already exists
ERROR:  relation "trips_car_id_index" already exists
ERROR:  relation "trips_end_address_id_index" already exists
ERROR:  relation "trips_start_address_id_index" already exists
ERROR:  relation "updates_car_id_index" already exists
ERROR:  constraint "cars_settings_id_fkey" for relation "cars" already exists
ERROR:  constraint "charges_charging_process_id_fkey" for relation "charges" already exists
ERROR:  constraint "charging_processes_address_id_fkey" for relation "charging_processes" already exists
ERROR:  constraint "charging_processes_car_id_fkey" for relation "charging_processes" already exists
ERROR:  constraint "charging_processes_geofence_id_fkey" for relation "charging_processes" already exists
ERROR:  constraint "charging_processes_position_id_fkey" for relation "charging_processes" already exists
ERROR:  constraint "drives_car_id_fkey" for relation "drives" already exists
ERROR:  constraint "drives_end_address_id_fkey" for relation "drives" already exists
ERROR:  constraint "drives_end_geofence_id_fkey" for relation "drives" already exists
ERROR:  constraint "drives_end_position_id_fkey" for relation "drives" already exists
ERROR:  constraint "drives_start_address_id_fkey" for relation "drives" already exists
ERROR:  constraint "drives_start_geofence_id_fkey" for relation "drives" already exists
ERROR:  constraint "drives_start_position_id_fkey" for relation "drives" already exists
ERROR:  constraint "positions_car_id_fkey" for relation "positions" already exists
ERROR:  constraint "positions_drive_id_fkey" for relation "positions" already exists
ERROR:  constraint "states_car_id_fkey" for relation "states" already exists
ERROR:  constraint "updates_car_id_fkey" for relation "updates" already exists

**Comment by CrowMash** [comment, marked as answer]
Did you run the requests detailed [here](https://docs.teslamate.org/docs/maintenance/backup_restore)?

# Stop the teslamate container to avoid write conflicts
docker-compose stop teslamate

# Drop existing data and reinitialize
docker-compose exec -T database psql -U teslamate << .
drop schema public cascade;
create schema public;
create extension cube;
create extension earthdistance;
CREATE OR REPLACE FUNCTION public.ll_to_earth(float8, float8)
    RETURNS public.earth
    LANGUAGE SQL
    IMMUTABLE STRICT
    PARALLEL SAFE
    AS 'SELECT public.cube(public.cube(public.cube(public.earth()*cos(radians(\$1))*cos(radians(\$2))),public.earth()*cos(radians(\$1))*sin(radians(\$2))),public.earth()*sin(radians(\$1)))::public.earth';
.

# Restore
docker-compose exec -T database psql -U teslamate -d teslamate < teslamate.bck

# Restart the teslamate container
docker-compose start teslamate

**Comment by pochacho99** [comment]
I'm having trouble running the "Drop existing data and reinitialize" command in the windows powershell, any advice?

________________________________

**Reply by CrowMash** [reply]
What kind of trouble? Do you have any error log to share?

Also, you have to make sure you're running powershell inside the same directory as your docker-compose file.

**Reply by pochacho99** [reply]
Thanks, i figured out how to run the reinitialize command and got my data back

**Reply by PoppyS33d** [reply]
I'm having the same issue, and I cannot drop the existing data. MacOS, in Terminal. Commands are just not found (Drop or Create), for example:

DROP SCHEMA public CASCADE;                           
zsh: command not found: DROP

Any further tips? Thank you.

**Reply by brianmay** [reply]
The commands need to be entered into a `psql` that is connected to the database. Not your shell.

**Reply by PoppyS33d** [reply]
Thanks for the reply. I have absolutely no idea how to do that on MacOS in Terminal, unfortunately.

**Reply by PoppyS33d** [reply]
In addition, I don't understand why this is even happening. I was just trying to restore my TeslaMate backup to a completely new system, and get all the errors listed above. There's no guidance on the GitHub for Mac users, it seems, on how to fix this, and it's way above my pay grade, unfortunately. I just do not know how to fix the issue.

**Reply by JakobLichterfeld** [reply]
> In addition, I don't understand why this is even happening.

You didn't follow https://docs.teslamate.org/docs/maintenance/backup_restore

> There's no guidance on the GitHub for Mac users,

MacOS is not a Server OS and not supported.

**Reply by PoppyS33d** [reply]
Respectfully, I exactly followed the instructions on that page.

**Reply by brianmay** [reply]
On MacOS you enter the same commands as anywhere else.

But based on what you have copied above you have not followed the instructions correctly. https://github.com/teslamate-org/teslamate/discussions/2972#discussioncomment-13868359

We do not aim to make everybody database experts. And unfortunately, operations like backup/restore of databases often tend to end up being non-trivial. But regardless we try to document common processes as best as possible.

Maybe the instructions as is are confusing. It uses a heredoc. That is indicated by the `<< .` - People may not understand what a heredoc is. Here is just one out of many tutorials: https://linuxize.com/post/bash-heredoc/

Possibly might be easier without the heredoc. "Type this command to get the psql prompt. Then type in these commands at the psql prompt. Then type in this to exit to psql prompt."

Or maybe what we need is is a mini-video that shows when and what you have to type, and what the expected responses are. If anybody has the time...

**Reply by PoppyS33d** [reply]
I think maybe we’re on different pages here. That’s alright. It’s working now and that’s all that matter, thanks!

**Comment by cwanja** [comment]
Do you have TeslaMate installed via Docker? If so, the comments are the same regardless of host OS. You simply use terminal and execute the command. On Jul 23, 2025, at 5:27 PM, PoppyS33d ***@***.***> wrote:﻿
In addition, I don't understand why this is even happening. I was just trying to restore my TeslaMate backup to a completely new system, and get all the errors listed above. There's no guidance on the GitHub for Mac users, it seems, on how to fix this, and it's way above my pay grade, unfortunately. I just do not know how to fix the issue.

**Reply by PoppyS33d** [reply]
Thanks for the reply. I did get this running using SQL with a friend's help, but I still now have my original issue (which is why I reinstalled TeslaMate in the first place, so back at square 1!). 

Here's the link to the new issue:

https://github.com/teslamate-org/teslamate/discussions/4851
